raft/README.md

```diff
@@ -2,4 +2,72 @@
 
 Raft implements distributed consensus in a more modular way than Paxos by using a strong leader and separating the algorithm into four modules that are not fully independent but intersect minimally: **leader election**, **log replication**, **snapshotting**, and **membership changes**. The last two were not studied in this course.
 
+## Server States
+(1) **Leader**: handles all client interactions and log replication  
+(2) **Follower**: completely passive, follow instructions of leader  
+(3) **Candidate**: a server that is vying to become the new leader 
 
+![state machine](/raft/state_machine.png)   
+
+* Every server maintains a **term** value, which is incremented when a follower turns into a candidate
+* Servers start as followers
+
+## Leader Behavior 
+* Leaders send **heartbeats** (empty AppendEntries RPCs) to maintain authority over followers
+* If an **electionTimeout** (150-300ms) passes with no heartbeat, the follower assumes the leader has crashed and starts a new election.
+
```

Comment:
missing leader behavior: send proposals (nonempty appendentries). Can only send proposals from current term bc of figure 8